### PR TITLE
J2735 2024 TIM Pipeline Improvements

### PR DESCRIPTION
**Pull request by drewjj**
# PR Details
## Description

Includes a complete overhaul of the TIM message pipeline for the UDP endpoint, log ingestion endpoint and TIM deposit REST endpoint.
- New J2735 2024 POJOs from the jpo-asn-pojos project have been incorporated into the TIM pipeline and all output now conforms 100% to the J2735 2024 as defined by the ASN.
- Unused and redundant POJO Kafka topics and classes have been removed.
- The TIM deposit procedure occurs in a slightly different order since the broadcast topics are no longer written to.
- Unit tests have been updated to conform to the new data formats and pipeline. Some of the approval tests required recreating the input to be compatible with the refactor.
- The flowchart has been updated to include the current state of the TIM pipeline.

## Related Issue

N/A

## How Has This Been Tested?

This has been tested locally in a single instance environment through the following methods:
- UDP endpoint ingestion
- rxMsg file log offload
- /tim deposit endpoint with signed and unsigned
- Unit tests were utilized to ensure known corner cases were tested properly

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [x] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
[ODE Contributing Guide](https://github.com/usdot-jpo-ode/jpo-ode/blob/develop/docs/contributing_guide.md) 
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/traveler/TimDepositController.java [hunk 1]
```diff
@@ -223,29 +229,10 @@ public synchronized ResponseEntity<String> depositTim(String jsonString, Request
...
-        odeTimData);
-
-    String obfuscatedTimData = TimTransmogrifier.obfuscateRsuPassword(odeTimData.toJson());
-    kafkaTemplate.send(jsonTopics.getTimBroadcast(), serialIdJ2735.getStreamId(),
-        obfuscatedTimData);
-
-    // Now that the message has been published to OdeBroadcastTim topic, it should
-    // be changed to J2735BroadcastTim serialId
+    // Set a unique serial ID for the ODE TIM message
+    SerialId serialIdJ2735 = new SerialId();
```
[the remark this answers is not in the document]
Comment (drewjj) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/traveler/TimDepositController.java:233]: This actually was never functioning this way and I really am not sure if there is justification for doing it either way.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/tim/TimReceiver.java [hunk 2]
```diff
@@ -43,19 +43,15 @@ public void run() {
...
         socket.receive(packet);
         if (packet.getLength() > 0) {
-
-          var tim = UdpHexDecoder.buildTimFromPacket(packet);
-          var timJson = tim.toJson();
-          if (timJson != null) {
-            timPublisher.send(publishTopic, tim.getMetadata().getSerialId().toString(), timJson);
+          String timData = UdpHexDecoder.buildJsonTimFromPacket(packet);
+          if (timData != null) {
+            timPublisher.send(publishTopic, timData);
```
[the remark this answers is not in the document]
Comment (drewjj) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/tim/TimReceiver.java:54]: Same logic as above.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/traveler/TimDepositController.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (drewjj) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/traveler/TimDepositController.java]: I have decided to rethink this in case your concern is the repeat of creating new SerialId objects. It is being created in the constructor and every deposit will make it increment at the start of the deposit. This way we don't have to worry about the number of many early returns that this function can go down.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/traveler/TimDepositController.java [hunk 3]
```diff
@@ -131,11 +132,13 @@ public TimDepositController(Asn1CoderTopics asn1CoderTopics, PojoTopics pojoTopi
...
    * @param jsonString The value of the JSON message
-   * @param verb       The HTTP verb being requested
+   * @param verb The HTTP verb being requested
    *
    * @return The request completion status
+   * @throws JsonProcessingException if there is an error processing the JSON
    */
-  public synchronized ResponseEntity<String> depositTim(String jsonString, RequestVerb verb) {
+  public synchronized ResponseEntity<String> depositTim(String jsonString, RequestVerb verb)
+      throws JsonProcessingException {
```
[the remark this answers is not in the document]
Comment (drewjj) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/traveler/TimDepositController.java]: Understood. I have addressed this. I believe TimDepositController is the only controller endpoints that do this.

File: jpo-ode-plugins/src/main/java/us/dot/its/jpo/ode/plugin/ServiceRequest.java [hunk 4]
```diff
@@ -96,9 +99,11 @@ public boolean equals(Object obj) {
 
   }
 
-  private static final long serialVersionUID = 1L;
+   // private static final long serialVersionUID = 1L;
    private OdeInternal ode;
    private SituationDataWarehouse.SDW sdw;
+   @JacksonXmlProperty(localName = "rsus")
+   @JacksonXmlElementWrapper(useWrapping = true, localName = "rsus")
```
[the remark this answers is not in the document]
Comment (drewjj) [reply jpo-ode-plugins/src/main/java/us/dot/its/jpo/ode/plugin/ServiceRequest.java]: This was actually remaining from some testing I was doing with deserialization but I left it thinking it was required after I had found the actual solution. I have now retested it and it isn't required.

File: jpo-ode-plugins/src/main/java/us/dot/its/jpo/ode/plugin/ServiceRequest.java [hunk 5]
```diff
@@ -96,9 +99,11 @@ public boolean equals(Object obj) {
 
   }
 
-  private static final long serialVersionUID = 1L;
+   // private static final long serialVersionUID = 1L;
```
[the remark this answers is not in the document]
Comment (drewjj) [reply jpo-ode-plugins/src/main/java/us/dot/its/jpo/ode/plugin/ServiceRequest.java]: I uncommented this.

File: jpo-ode-svcs/src/test/java/us/dot/its/jpo/ode/traveler/TimDepositControllerTest.java [hunk 6]
```diff
@@ -1081,24 +660,69 @@ void testSuccessfulRsuMessageReturnsSuccessMessagePost() throws IOException {
...
+    Assertions.assertTrue(xml.contains("<rsuIndex>1</rsuIndex>"));
+    Assertions.assertTrue(xml.contains("<rsuUsername>admin</rsuUsername>"));
+    Assertions.assertTrue(xml.contains("<rsuPassword>password</rsuPassword>"));
+    Assertions.assertTrue(xml.contains("<snmpProtocol>NTCIP1218</snmpProtocol>"));
+    Assertions.assertTrue(xml.contains("</rsus></rsus>"));
+
+    //ServiceRequest deserialized = xmlMapper.readValue(xml, ServiceRequest.class);
+    ObjectNode consumed = xmlMapper.readValue(xml, ObjectNode.class);
+    String xmlBack = xmlMapper.writeValueAsString(consumed);
+    ServiceRequest deserialized = xmlMapper.readValue(xmlBack, ServiceRequest.class);
```
[the remark this answers is not in the document]
Comment (drewjj) [reply jpo-ode-svcs/src/test/java/us/dot/its/jpo/ode/traveler/TimDepositControllerTest.java]: This was the issue I was troubleshooting with the annotations. I honestly can't tell you why. It just doesn't work. I drained 2 hours into this.

File: jpo-ode-svcs/src/test/java/us/dot/its/jpo/ode/traveler/TimDepositControllerTest.java [hunk 7]
```diff
@@ -1081,24 +660,69 @@ void testSuccessfulRsuMessageReturnsSuccessMessagePost() throws IOException {
...
-    pojoTimBroadcastConsumer.close();
-    jsonTimBroadcastConsumer.close();
-    jsonJ2735TimBroadcastConsumer.close();
     jsonTimConsumer.close();
     asn1CoderEncoderInputConsumer.close();
     DateTimeUtils.setClock(prevClock);
   }
 
+@Test
+void testServiceRequestWithRsusSerializationToXml() throws Exception {
```
[the remark this answers is not in the document]
Comment (drewjj) [reply jpo-ode-svcs/src/test/java/us/dot/its/jpo/ode/traveler/TimDepositControllerTest.java]: I have created the ServiceRequestTest test class and moved it there.

File: jpo-ode-svcs/src/test/resources/application.yaml [hunk 8]
```diff
@@ -43,7 +43,7 @@ ode:
       value-serializer: "org.apache.kafka.common.serialization.ByteArraySerializer"
       partitioner-class: "org.apache.kafka.clients.producer.internals.DefaultPartitioner"
     disabled-topics:
-      - "topic.OdeTimBroadcastPojo"
+      - ""
```
[the remark this answers is not in the document]
Comment (drewjj) [reply jpo-ode-svcs/src/test/resources/application.yaml:46]: This was a bit lazy to be honest. The current disable topic support seems to break when there isn't at least one disabled topic. Since I was in serious crunch mode I chose to go this route.

File: jpo-ode-svcs/src/test/java/us/dot/its/jpo/ode/traveler/TimDepositControllerTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (mcook42) [reply jpo-ode-svcs/src/test/java/us/dot/its/jpo/ode/traveler/TimDepositControllerTest.java]: since it's a test, I'm ok moving forward with this workaround

File: jpo-ode-plugins/src/main/java/us/dot/its/jpo/ode/plugin/ServiceRequest.java [hunk 9]
```diff
@@ -96,7 +99,7 @@ public boolean equals(Object obj) {
 
   }
 
-  private static final long serialVersionUID = 1L;
+   private static final long serialVersionUID = 1L;
```
[the remark this answers is not in the document]
Comment (drewjj) [reply jpo-ode-plugins/src/main/java/us/dot/its/jpo/ode/plugin/ServiceRequest.java]: Addressed

**Review by mcook42 — APPROVED** [review]

**Review by dmccoystephenson — CHANGES_REQUESTED** [review]
It's nice to get the TIM pipeline cleaned up like this!

I had some non-blocking comments and I ran into some test failures:
image

File: jpo-ode-plugins/src/main/java/us/dot/its/jpo/ode/plugin/ServiceRequest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (drewjj) [reply jpo-ode-plugins/src/main/java/us/dot/its/jpo/ode/plugin/ServiceRequest.java]: This is no longer commented out

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/traveler/TimDepositController.java [hunk 10]
```diff
@@ -180,15 +178,23 @@ public synchronized ResponseEntity<String> depositTim(String jsonString, Request
           .body(JsonUtils.jsonKeyValue(ERRSTR, errMsg));
     }
 
-    // Add metadata to message and publish to kafka
+    // Short circuit
+    // If the TIM has no RSU/SNMP or SDW structures, return a warning
+    if ((request.getRsus() == null || request.getSnmp() == null) && request.getSdw() == null) {
+      String warningMsg = "Warning: TIM contains no RSU, SNMP, or SDW fields."
+          + " Message only published to broadcast streams.";
```
[the remark this answers is not in the document]
Comment (dmccoystephenson) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/traveler/TimDepositController.java:185]: That's right, I do remember that. This warning message will be misleading then and should probably be modified to not include "Message only published to broadcast streams"

[the remark this answers is not in the document]
Comment (drewjj) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/traveler/TimDepositController.java:185]: You are right, I will remove that.

[the remark this answers is not in the document]
Comment (drewjj) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/traveler/TimDepositController.java]: Addressed

**Comment by drewjj** [comment]
Note: This PR is now too big with all of the deletions for checkstyle. Please ignore the checkstyle error

**Comment by drewjj** [comment]
Ready for review again!
